main-command/src/main/scala/sbt/internal/server/Server.scala

```diff
@@ -54,15 +54,17 @@ private[sbt] object Server {
       val serverThread = new Thread("sbt-socket-server") {
         override def run(): Unit = {
           Try {
-            ErrorHandling.translate(s"server failed to start on ${connection.shortName}. ") {
-              connection.connectionType match {
-                case ConnectionType.Local if isWindows =>
-                  new NGWin32NamedPipeServerSocket(pipeName)
-                case ConnectionType.Local =>
-                  prepareSocketfile()
-                  new NGUnixDomainServerSocket(socketfile.getAbsolutePath)
-                case ConnectionType.Tcp => new ServerSocket(port, 50, InetAddress.getByName(host))
-              }
+            connection.connectionType match {
+              case ConnectionType.Local if isWindows =>
+                // Named pipe already has an exclusive lock.
+                addServerError(new NGWin32NamedPipeServerSocket(pipeName))
+              case ConnectionType.Local =>
+                tryClient(new NGUnixDomainSocket(socketfile.getAbsolutePath))
+                prepareSocketfile()
+                addServerError(new NGUnixDomainServerSocket(socketfile.getAbsolutePath))
```

Comment:
For both `ConnectionType.Local ` and for `ConnectionType.Tcp`, why do the tried sockets not match the used sockets?

Reply:
1. `tryClient(...)` pretends to be a client and hits the server's URL as if it's there. Thus if the socket fails it's a success (no server yet), and if the socket connects it's a failure (server is already up).
2. The second socket is the real server socket.

Reply:
Ah right, gotcha.